src/extension_commands.ml

```diff
@@ -534,6 +534,143 @@ module Copy_type_under_cursor = struct
     command Extension_consts.Commands.copy_type_under_cursor handler
 end
 
+module Search_by_type = struct
+  let extension_name = "Search By Type"
+
+  let ocaml_lsp_doesnt_support_search_by_type instance ocaml_lsp =
+    match
+      Ocaml_lsp.is_version_up_to_date
+        ocaml_lsp
+        (Extension_instance.ocaml_version_exn instance)
+    with
+    | Ok () -> ()
+    | Error (`Msg msg) ->
+      show_message
+        `Warn
+        "The installed version of `ocamllsp` does not support type search. %s"
+        msg
+
+  let rec remove_duplicates = function
+    | (a : Custom_requests.Type_search.type_search_result) :: (b :: _ as t) ->
+      if String.equal a.name b.name && a.cost = b.cost then remove_duplicates t
+      else a :: remove_duplicates t
+    | unique -> unique
+
+  let get_search_results ~query ~limit ~with_doc ~position text_editor client =
+    let open Promise.Syntax in
+    let doc = TextEditor.document text_editor in
+    let uri = TextDocument.uri doc in
+    let* res =
+      Custom_requests.(
+        send_request
+          client
+          Type_search.request
+          (Type_search.make ~uri ~position ~limit ~query ~with_doc ()))
+    in
+    match res with
+    | [] ->
+      show_message `Info "Empty results";
+      [] |> Promise.return
+    | res -> remove_duplicates res |> Promise.return
```

Comment:
This function is purely utilitary, we should always remove duplicates and handle the empty case at it's call-site.

Additionally you can use `let+` so that you don't need all these `Promise.return`